pkg/builder/endpoints.go

```diff
@@ -0,0 +1,294 @@
+	// Route53: https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-values-geo.html
+	// Google: https://cloud.google.com/compute/docs/regions-zones
+	Geo string
+
+	// IsDefaultGeo specifies if this is the default geo for providers that support setting a default catch all geo endpoint such as Route53
+	IsDefaultGeo bool
+}
+
+// NewEndpointsBuilder returns a new endpoints builder
+func NewEndpointsBuilder(target Target, hostname string) *EndpointsBuilder {
+	return &EndpointsBuilder{
+		target:   target,
+		hostname: hostname,
+	}
+}
+
+// WithLoadBalancing provides builder with necessary parameters to generate a load-balancing set of endpoints.
+// If not used, the builder will provide a simple set of endpoints
+func (builder *EndpointsBuilder) WithLoadBalancing(id string, loadbalancing *LoadBalancing) *EndpointsBuilder {
+	builder.id = id
+	builder.loadBalancing = loadbalancing
+	return builder
+}
+
+// WithLoadBalancingFor performs identically to WithLoadBalancing but without the need to parse params in LoadBalancing struct
+func (builder *EndpointsBuilder) WithLoadBalancingFor(id string, weight int, geo string, isDefaultGeo bool) *EndpointsBuilder {
+	return builder.WithLoadBalancing(id, &LoadBalancing{
+		Weight:       weight,
+		Geo:          geo,
+		IsDefaultGeo: isDefaultGeo,
+	})
+}
+
+// Build returns a list of endpoints created based on the current configuration of the builder
+func (builder *EndpointsBuilder) Build() ([]*externaldns.Endpoint, error) {
+	if err := builder.Validate(); err != nil {
+		return nil, err
+	}
+
+	var endpoints []*externaldns.Endpoint
+
+	// no load-balancing provided, inferring simple strategy
+	if builder.loadBalancing == nil {
+		endpoints = builder.getSimpleEndpoints()
+	} else {
+		// load-balancing present, inferring load-balanced strategy
+		endpoints = builder.getLoadBalancedEndpoints()
+	}
+
+	sort.Slice(endpoints, func(i, j int) bool {
+		return getSetID(endpoints[i]) < getSetID(endpoints[j])
+	})
+	return endpoints, nil
+}
+
+// Validate set of parameters passed inside the builder.
+// Does not ensure validity of values, but checks for a correct set of inputs.
+// e.g. will not check for invalid address type, but ensure they are not nil
+func (builder *EndpointsBuilder) Validate() error {
+	if matched, err := regexp.MatchString(HostnameRegex, builder.hostname); !matched {
+		// This only possible if HostnameRegex is modified.
+		// Leave it here as a precaution
+		if err != nil {
+			return fmt.Errorf("error parsing regexp to match hostname: %w", err)
+		}
+		return fmt.Errorf("invalid hostname")
+	}
+
+	if builder.target == nil {
+		return fmt.Errorf("must provide target")
+	}
+
+	if builder.target.GetAddresses() == nil {
+		return fmt.Errorf("must provide addresses")
+	}
+
+	// following only applicable for load-balancing strategy
+	if builder.loadBalancing != nil {
+		// id must not be an empty string
+		if builder.id == "" {
+			return fmt.Errorf("ID is required")
+		}
```

Comment:
This, id,  to me should just be made part of the loadBalancing struct since its only needed for loadbalanced endpoints

Reply:
sure, why not. Moved